### PR TITLE
Expand parquet crate overview doc

### PR DESCRIPTION
**Pull request by mmaitre314**
# Which issue does this PR close?

This is a doc-only update converting the discussions in #5064 and #2394 into docs. It closes neither of these PRs -- code fixes are needed for that.

# Rationale for this change
 
The change clarifies which APIs are available in the parquet crate, what their intents are, how they relate to one-another, and which should be used when.

# Are there any user-facing changes?

Doc update

### CODE REVIEW

**Comment by tustvold** [comment]
This looks fantastic, thank you. I'm not at a computer currently but I'll double check dull things like formatting when I am, but from a quick read through the content looks :ok_hand:

File: parquet/README.md [hunk 1]
```diff
@@ -71,14 +71,6 @@ The `parquet` crate provides the following features which may be enabled in your
 - [x] Predicate pushdown
 - [x] Parquet format 4.0.0 support
 
-## Support for `wasm32-unknown-unknown` target
```
Comment (tustvold) [inline parquet/README.md:74]: This is out-dated and no longer true so I took the opportunity to remove it

File: parquet/Cargo.toml [hunk 2]
```diff
@@ -115,11 +115,6 @@ name = "async_read_parquet"
 required-features = ["arrow", "async"]
 path = "./examples/async_read_parquet.rs"
 
-[[example]]
```
Comment (tustvold) [inline parquet/Cargo.toml:117]: I opted to just include the doctest as an example

File: parquet/src/arrow/async_reader/store.rs [hunk 3]
```diff
@@ -28,42 +28,30 @@ use crate::arrow::async_reader::{AsyncFileReader, MetadataLoader};
...
+/// # use object_store::ObjectStore;
+/// # use object_store::path::Path;
+/// # use parquet::arrow::async_reader::ParquetObjectReader;
+/// # use parquet::arrow::ParquetRecordBatchStreamBuilder;
+/// # use parquet::schema::printer::print_parquet_metadata;
+/// # async fn main() {
+/// // Populate configuration from environment
+/// let storage_container = Arc::new(MicrosoftAzureBuilder::from_env().build()?);
+/// let location = Path::from("path/to/blob.parquet");
+/// let meta = storage_container.head(&location).await?;
```
Comment (tustvold) [inline parquet/src/arrow/async_reader/store.rs:46]: Head is a more efficient way to obtain the ObjectMeta, as it avoids fetching any data

File: parquet/src/arrow/async_reader/store.rs [hunk 4]
```diff
@@ -28,42 +28,30 @@ use crate::arrow::async_reader::{AsyncFileReader, MetadataLoader};
...
+/// # use std::sync::Arc;
+/// # use object_store::azure::MicrosoftAzureBuilder;
+/// # use object_store::ObjectStore;
+/// # use object_store::path::Path;
+/// # use parquet::arrow::async_reader::ParquetObjectReader;
+/// # use parquet::arrow::ParquetRecordBatchStreamBuilder;
+/// # use parquet::schema::printer::print_parquet_metadata;
+/// # async fn main() {
+/// // Populate configuration from environment
+/// let storage_container = Arc::new(MicrosoftAzureBuilder::from_env().build()?);
```
Comment (tustvold) [inline parquet/src/arrow/async_reader/store.rs:44]: I tweaked this to use from_env to cut down on boilerplate

File: parquet/src/file/mod.rs [hunk 5]
```diff
@@ -15,7 +15,7 @@
 // specific language governing permissions and limitations
 // under the License.
 
-//! Main entrypoint for working with Parquet API.
+//! Low level APIs for reading raw parquet data.
```
Comment (tustvold) [inline parquet/src/file/mod.rs:18]: I did not realise this was here, this was deeply misleading :smile:

File: parquet/src/lib.rs [hunk 6]
```diff
@@ -15,70 +15,67 @@
...
-[`arrow::async_reader`] and [`arrow::async_writer`]. These APIs are more advanced and
-require the `async` feature. Within this module,
-[`ParquetObjectReader`](arrow::async_reader::ParquetObjectReader)
-enables connecting directly to the Cloud Provider storage services of AWS, Azure, GCP
-and the likes via the [object_store](https://docs.rs/object_store/latest/object_store/)
-crate, enabling network-bandwidth optimizations via predicate and projection push-downs.
-
-*/
+//!
+//! This crate contains the official Native Rust implementation of
```
Comment (tustvold) [inline parquet/src/lib.rs:19]: I tweaked the copy slightly to be a bit more concise and easier to parse

File: parquet/src/lib.rs [hunk 7]
```diff
@@ -15,70 +15,67 @@
...
-read and written one row group at a time by calling respectively
-[`SerializedFileReader::get_row_group`](file::serialized_reader::SerializedFileReader)
- and [`SerializedFileWriter::next_row_group`](file::writer::SerializedFileWriter).
-Within each row group, columns are read and written one at a time using
-respectively [`ColumnReader`](column::reader::ColumnReader) and
-[`ColumnWriter`](column::writer::ColumnWriter). The [`mod@file`] module also allows
-reading files in a row-wise manner via
-[`SerializedFileReader::get_row_iter`](file::serialized_reader::SerializedFileReader).
-This is a convenience API which favors simplicity over performance and completeness.
-It is not recommended for production use.
```
Comment (tustvold) [inline parquet/src/lib.rs]: I opted to move this into a caveats section on RowIter and expand a bit more on the why

File: parquet/src/lib.rs [hunk 8]
```diff
@@ -15,70 +15,67 @@
...
-are stored as integers (physical type). This crate exposes both kinds of types.
-
-For more details about the Parquet format, see the
-[Parquet spec](https://github.com/apache/parquet-format/blob/master/README.md#file-format).
-
-## APIs
-
-This crate exposes both low-level and high-level APIs, organized as follows:
-
-1. The [`arrow`] module reads and writes Parquet data to/from Arrow
```
Comment (tustvold) [inline parquet/src/lib.rs]: I split these into headed sections, as they felt a bit meaty to be bullets

File: parquet/src/lib.rs [hunk 9]
```diff
@@ -15,70 +15,67 @@
...
-`RecordBatch`es. This is the recommended high-level API. It allows leveraging
-the wide range of data transforms provided by the
-[arrow](https://docs.rs/arrow/latest/arrow/index.html) crate and by the ecosystem
-of libraries and services using Arrow as a interop format.
-
-2. The [`mod@file`] module allows reading and writing Parquet files without taking a
-dependency on Arrow. This is the recommended low-level API. Parquet files are
-read and written one row group at a time by calling respectively
-[`SerializedFileReader::get_row_group`](file::serialized_reader::SerializedFileReader)
```
Comment (tustvold) [inline parquet/src/lib.rs]: I opted to de-emphasise these APIs, as they're low-level and I think most people should just steer clear of them

**Review by tustvold — APPROVED** [review]
Thank you so much for this, I moved some things around but what you wrote was already very good, I was just inspired to make some further changes

**Comment by tustvold** [comment]
As I've made some non-trivial modifications I intend to leave this open for until Monday, to give other reviewers time to comment

**Comment by mmaitre314** [comment]
> As I've made some non-trivial modifications I intend to leave this open for until Monday, to give other reviewers time to comment

Thanks! The update looks good to me. The one piece I would keep out would be the use of `from_env()`, because it changes the scenario. It looks like `from_env()` assumes there is a single Storage Account with a single Container and only the Blob Path changes each time, which is not very typical in data pipelines where datasets may come from different teams and may be partitioned to avoid throttling, to implement different access control lists, or abide by different regional regulatory requirements. The one boiler-plate line is `.with_use_azure_cli(true)`. Were an official Azure SDK available, this would not be there and instead `DefaultAzureCredential` would be used to handle auth transparently and have the same code run during local development and remote deployment.

**Comment by tustvold** [comment]
My intention was to focus on showing how to configure integration with object_store, not provide an exhaustive list of all the various ways one might configure it. `from_env` provides a minimal viable baseline that people can then override as they see fit within their applications

**Pull request merged**